scarlet/component.py

```diff
@@ -365,370 +176,53 @@ class CubeComponent(Component):
-        return self
-
-    def __getitem__(self, coord):
-        """Access node in the tree.
-
-        Parameters
-        ----------
-        coords: int or tuple of ints
-
-        Returns
-        -------
-        `~scarlet.component.Component` or `~scarlet.component.ComponentTree`
-        """
-        if isinstance(coord, (tuple, list)):
-            if len(coord) > 1:
-                return self._tree[coord[0]].__getitem__(coord[1:])
-            else:
-                return self._tree[coord[0]]
-        elif isinstance(coord, int):
-            return self._tree[coord]
-        else:
-            raise NotImplementedError("coord needs to be index or list of indices")
-
-    def __getstate__(self):
-        # needed for pickling to understand what to save
-        return (self._tree,)
+    def get_model(self, *parameters, frame=None):
+        model = self.get_parameter(0, *parameters)
 
-    def __setstate__(self, state):
-        self._tree = state[0]
-        self._tree_to_components()
+        if frame is not None:
+            model = self.model_to_frame(frame, model)
+        return model
 
-    def model_to_frame(self, frame=None, model=None):
-        """Project a model into a frame
 
+class CombinedComponent(Component):
+    def __init__(self, components, operation="add", check_boxes=True):
 
-        Parameters
-        ----------
-        model: array
-            Image of the model to project.
-            This must be the same shape as `self.bbox`.
-            If `model` is `None` then `self.get_model()` is used.
-        frame: `~scarlet.frame.Frame`
-            The frame to project the model into.
-            If `frame` is `None` then the model is projected
-            into `self.model_frame`.
+        assert len(components)
+        frame = components[0].frame
+        box = components[0].bbox
+        # all children need to have the same bbox for simple autogradable combinations
+        for c in components:
+            assert isinstance(c, Component)
+            assert c.frame is frame
+            if check_boxes:
+                assert c.bbox == box
 
-        Returns
-        -------
-        projected_model: array
-            (Channels, Height, Width) image of the model
-        """
-        # Use the current model by default
-        if model is None:
-            model = self.get_model()
-        # Use the full model frame by default
-        if frame is None:
-            frame = self.model_frame
+        super().__init__(frame, children=components, bbox=box)
 
-        if hasattr(frame, "dtype"):
-            dtype = frame.dtype
-        else:
-            dtype = self.model_frame.dtype
+        assert operation in ["add", "multiply"]
+        self.operation = operation
 
-        frame_slices, model_slices = overlapped_slices(frame, self.bbox)
+    def get_model(self, *parameters, frame=None):
+        # boxed models
+        models = self.get_models_of_children(*parameters, frame=None)
```

Comment:
Your's is the proposal that brings about the potential race condition because we have to declare `_add_model` as a differentiable function. I avoid that by using simple (autograd) numpy operations `+` or `*`. But that requires that boxes have same shape

Reply:
Yes, I realize that. What I meant is that if you pass `frame=self.bbox` then it will pad the children with zeros so that they are the same shape as the combined model. So if, for example, we have a model that has a really small bulge and a really large disk, in your model (and I think the way that the code works now) the bulge parameter will be an array with a much larger number of entries than we actually need to fit. So to fit with this scheme we are wasting a lot of operations. However, if you just project it into the frame for the `CombinedComponent` even though we know padding by zeros is slow, if the size difference between the bulge and the disk is large enough then overall we save a lot of cpu cycles.

A better example might be a dust model. If we have an extended galaxy with dust that obscures more distant galaxies behind it, the dust component will have its host galaxy as well as the other galaxies that it overlaps as children, which will all be much smaller than the closer dusty galaxy. So to make this work in the current code you have to make all of the overlapping galaxies the same size, which requires joint initialization and much larger (non-centered) morphologies, which we currently don't allow. However if you just allow the `CombinedComponent` to project them to its frame then we don't have to worry about any of that. It won't be as efficient as the slicing mechanism, but it should still be faster.

Reply:
I can get the old padding code back, sure, but I'm surprised that you want to see it reinstated here.

As for dust, I think we should try only obscuring on galaxy, namely its host, and not also the background. Small steps...

Reply:
I know it's slower, but the code is already implemented, you just have to change `frame` here and add a check in `model_to_frame` that if `frame==bbox` to just return the model. So if the user does put all of the sources in the same `Box` then it won't really cost anything.

Reply:
Re dust: I think about the blend that we first started thinking is poorly modeled due to dust. There are a number of smaller, more distant galaxies that are blended with it. So if we can model the dust for the foreground galaxy then we should, otherwise the colors->redshift of the background galaxies will be off.

Sure, that might not be where we start, but we should definitely try to get there...

Reply:
Then you have to know which of the galaxies are behind. If you get it wrong it will affect all the others.

Reply:
Agreed. I'm not saying that I know how to solve this now, just that it's something that we might want the option to do in the future, so there's no harm in putting it in now.

Reply:
Added in next commit